<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>S3: Distances, clustering • CSHstats</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="../deps/MathJax-3.2.2/tex-chtml.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="S3: Distances, clustering">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">CSHstats</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.35</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/S1_probability.html">S1_probability: Definitions and exercises</a></li>
    <li><a class="dropdown-item" href="../articles/S1b_hyper.html">S1a: A look at the hypergeometric distribution</a></li>
    <li><a class="dropdown-item" href="../articles/S2_EDA.html">S2_EDA: Exploratory data analysis and visualization</a></li>
    <li><a class="dropdown-item" href="../articles/S2a_PCAelem.html">S2a: Elements of PCA</a></li>
    <li><a class="dropdown-item" href="../articles/S3_distclust.html">S3: Distances, clustering</a></li>
    <li><a class="dropdown-item" href="../articles/S4_hypothesis_testing.html">S4_inference: Definitions and exercises</a></li>
    <li><a class="dropdown-item" href="../articles/S5_linear_models.html">S5_inference: FDR, linear models, and GLMs</a></li>
    <li><a class="dropdown-item" href="../articles/S5a_lmdiagnostics.html">S5a: Diagnostics of a differential gene expression exercise</a></li>
    <li><a class="dropdown-item" href="../articles/S5b_nonlin.html">S5b: hypothesis testing in nonlinear regression</a></li>
    <li><a class="dropdown-item" href="../articles/S6_hca_onto.html">S6 HCA, ontologies, shiny</a></li>
    <li><a class="dropdown-item" href="../articles/S7_cnn.html">S7 Neural net concepts and applications</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/vjcitn/CSHstats/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>S3: Distances, clustering</h1>
                        <h4 data-toc-skip class="author">Vincent J. Carey, stvjc at channing.harvard.edu</h4>
            
            <h4 data-toc-skip class="date">January 09, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/CSHstats/blob/HEAD/vignettes/S3_distclust.Rmd" class="external-link"><code>vignettes/S3_distclust.Rmd</code></a></small>
      <div class="d-none name"><code>S3_distclust.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="road-map">Road map<a class="anchor" aria-label="anchor" href="#road-map"></a>
</h2>
<ul>
<li>Concepts underlying cluster analysis
<ul>
<li>Distances in high-dimensional spaces</li>
<li>Criteria for agglomerative clustering</li>
</ul>
</li>
<li>Exploring a cluster analysis of copy-number aberration measures in 28 tumors from one patient
<ul>
<li>interactive heatmap</li>
<li>comparing Euclidean and correlation (1-cor) distances</li>
<li>comparing agglomeration methods</li>
<li>the silhouette as a measure of clustering adequacy</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="cluster-analysis-concepts">Cluster analysis concepts<a class="anchor" aria-label="anchor" href="#cluster-analysis-concepts"></a>
</h2>
<p>The need to identify groups of similar observations arises in many contexts – ultimately in the service of clarifying sources of variability, building power for statistical comparisons.</p>
<p>Hierarchical clustering of N multivariate observations is conducted by starting with N clusters and proceeding from M clusters to M-1 (M=N, …, 2) clusters by merging the two members separated by the least of all pairwise distances.</p>
<p><img src="agglomPic.jpg"></p>
<div class="section level3">
<h3 id="distances-in-high-dimensional-space">Distances in high-dimensional space<a class="anchor" aria-label="anchor" href="#distances-in-high-dimensional-space"></a>
</h3>
<p><img src="distances.jpg"></p>
</div>
<div class="section level3">
<h3 id="criteria-for-agglomerative-clustering">Criteria for agglomerative clustering<a class="anchor" aria-label="anchor" href="#criteria-for-agglomerative-clustering"></a>
</h3>
<p><img src="agglom.jpg"></p>
</div>
</div>
<div class="section level2">
<h2 id="application-inferring-steps-in-tumor-metastasis-in-a-breast-cancer-patient">Application: inferring steps in tumor metastasis in a breast cancer patient<a class="anchor" aria-label="anchor" href="#application-inferring-steps-in-tumor-metastasis-in-a-breast-cancer-patient"></a>
</h2>
<p>We’ll examine data distributed with a <a href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-021-00989-6" class="external-link">2021 Genome Biology paper</a> from the Gabor Marth lab.</p>
<div class="figure">
<img src="marthPatient.jpg" alt=""><p class="caption">Clinical sequence of interventions.</p>
</div>
<div class="figure">
<img src="marthEvolve.jpg" alt=""><p class="caption">Event sequence.</p>
</div>
<div class="section level3">
<h3 id="a-view-of-copy-number-aberrations-for-1mb-tiling">A view of copy number aberrations for 1Mb tiling<a class="anchor" aria-label="anchor" href="#a-view-of-copy-number-aberrations-for-1mb-tiling"></a>
</h3>
<p>28 tumors were sampled and sequenced in a rapid autopsy procedure. Copy number variation was assessed using <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5027494/" class="external-link">FACETS</a>.</p>
<p>The tissues from which tumors were taken are Br (Breast), Bo (Bone), Bn (Brain), Ln (Lung), Lv (Liver), Pa (Pancreas), Ly (Lymph nodes), Kd (Kidney)</p>
<p><a href="cnv-heatmap.html">A plotly-based visualization</a></p>
<p>The (vertical) ordering of tissues is chosen to exemplify certain similarities.</p>
<p>For example the block of blue on chr10 is seen for only three samples. This is an indication of a deletion.</p>
</div>
<div class="section level3">
<h3 id="a-cluster-analysis-proposed-in-support-of-the-evolutionary-map">A cluster analysis proposed in support of the evolutionary map<a class="anchor" aria-label="anchor" href="#a-cluster-analysis-proposed-in-support-of-the-evolutionary-map"></a>
</h3>
<p>This code is lightly modified from a script distributed at <a href="https://github.com/xiaomengh/tumor-evo-rapid-autopsy.git" class="external-link uri">https://github.com/xiaomengh/tumor-evo-rapid-autopsy.git</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span> <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">csamaDist</span><span class="op">)</span></span>
<span> <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bioDist</span><span class="op">)</span></span>
<span> <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://svn.r-project.org/R-packages/trunk/cluster/" class="external-link">cluster</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error in get(paste0(generic, ".", class), envir = get_method_env()) : </span></span>
<span><span class="co">##   object 'type_sum.accel' not found</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">cnv_log_R</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">=</span> <span class="va">cnv_log_R</span></span>
<span><span class="va">samples</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Ln7'</span>,<span class="st">'Ln9'</span>,<span class="st">'Ln1'</span>,<span class="st">'BrM'</span>,<span class="st">'BrP'</span>,</span>
<span>           <span class="st">'Ln11'</span>,<span class="st">'Ly2'</span>,<span class="st">'Ln3'</span>,</span>
<span>           <span class="st">'Bo3'</span>,<span class="st">'Ln10'</span>,<span class="st">'Bo1'</span>,<span class="st">'Ln8'</span>,<span class="st">'Lv3'</span>,<span class="st">'Ln5'</span>,<span class="st">'Bo2'</span>,<span class="st">'Bn2'</span>,<span class="st">'Bn1'</span>,<span class="st">'Bn3'</span>,<span class="st">'Bn4'</span>,<span class="st">'Ln2'</span>,</span>
<span>           <span class="st">'Ly1'</span>,<span class="st">'Ln6'</span>,</span>
<span>           <span class="st">'Kd1'</span>,<span class="st">'Ln4'</span>,<span class="st">'Lv4'</span>,<span class="st">'Lv2'</span>,<span class="st">'Lv1'</span>,<span class="st">'Pa1'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">=</span> <span class="va">samples</span></span>
<span><span class="va">d</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="va">data</span>, method<span class="op">=</span><span class="st">"euclidean"</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="va">d</span>, method<span class="op">=</span><span class="st">"average"</span><span class="op">)</span></span>
<span><span class="co"># the following line changes the order of the samples to produce the Fig.S3B but doesn't change the phylogenetic relationship</span></span>
<span><span class="va">fit</span><span class="op">$</span><span class="va">order</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">4</span>,<span class="fl">2</span>,<span class="fl">5</span>,<span class="fl">3</span>,<span class="fl">13</span>,<span class="fl">10</span>,<span class="fl">20</span>,<span class="fl">16</span>,<span class="fl">11</span>,<span class="fl">12</span>,<span class="fl">15</span>,<span class="fl">9</span>,<span class="fl">17</span>,<span class="fl">19</span>,<span class="fl">14</span>,<span class="fl">18</span>,<span class="fl">21</span>,<span class="fl">22</span>,<span class="fl">7</span>,<span class="fl">6</span>,<span class="fl">8</span>,<span class="fl">25</span>,<span class="fl">27</span>,<span class="fl">26</span>,<span class="fl">28</span>,<span class="fl">23</span>,<span class="fl">24</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<p><img src="S3_distclust_files/figure-html/doclus-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="drilling-down-on-the-clustering">Drilling down on the clustering<a class="anchor" aria-label="anchor" href="#drilling-down-on-the-clustering"></a>
</h3>
<div class="section level4">
<h4 id="comparing-euclidean-and-correlation-distances">Comparing Euclidean and Correlation distances<a class="anchor" aria-label="anchor" href="#comparing-euclidean-and-correlation-distances"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cd</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bioDist/man/cor.dist.html" class="external-link">cor.dist</a></span><span class="op">(</span><span class="va">cnv_log_R</span><span class="op">)</span> <span class="co"># from bioDist</span></span>
<span><span class="va">ed</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="va">cnv_log_R</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">ed</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">cd</span><span class="op">)</span>, xlab<span class="op">=</span><span class="st">"All pairwise Euclidean distances"</span>, ylab<span class="op">=</span><span class="st">"All pairwise correlation distances"</span><span class="op">)</span></span></code></pre></div>
<p><img src="S3_distclust_files/figure-html/dodists-1.png" width="700"></p>
<p>For a given correlation distance value, there can be wide variation in euclidean distance, and vice versa.</p>
<p>Open question: What distance metric is most relevant for biological interpretation of CNV?</p>
</div>
<div class="section level4">
<h4 id="a-pair-with-discrepant-correlation-and-euclidean-distance-values-over-entire-genome">A pair with discrepant correlation and euclidean distance values (over entire genome)<a class="anchor" aria-label="anchor" href="#a-pair-with-discrepant-correlation-and-euclidean-distance-values-over-entire-genome"></a>
</h4>
<p>We’ll have a look at the first 100Mb on chr1.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">cnv_log_R</span><span class="op">[</span><span class="st">"Ly1"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span>,pch<span class="op">=</span><span class="fl">19</span>, main<span class="op">=</span><span class="st">"chr1, first 100Mb"</span>, ylab<span class="op">=</span><span class="st">"FACETS CNV log R"</span>, xlab<span class="op">=</span><span class="st">"chr1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span><span class="va">cnv_log_R</span><span class="op">[</span><span class="st">"Ln1"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span>, col<span class="op">=</span><span class="st">"red"</span>,pch<span class="op">=</span><span class="fl">19</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="fl">60</span>, <span class="op">-</span><span class="fl">.5</span>, pch<span class="op">=</span><span class="fl">19</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span>, legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Ly1"</span>, <span class="st">"Ln1"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="S3_distclust_files/figure-html/lk1mb-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#cor(cnv_log_R["Ly1", 1:100], cnv_log_R["Ln1", 1:100])</span></span>
<span><span class="va">edist</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">(</span><span class="va">x</span><span class="op">-</span><span class="va">y</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">edist</span><span class="op">(</span><span class="va">cnv_log_R</span><span class="op">[</span><span class="st">"Ly1"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span>, <span class="va">cnv_log_R</span><span class="op">[</span><span class="st">"Ln1"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1.685992</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/jitter.html" class="external-link">jitter</a></span><span class="op">(</span><span class="va">cnv_log_R</span><span class="op">[</span><span class="st">"Ly1"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span><span class="op">)</span>, <span class="va">cnv_log_R</span><span class="op">[</span><span class="st">"Ln1"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span>, xlab<span class="op">=</span><span class="st">"Ly1"</span>, ylab<span class="op">=</span><span class="st">"Ln1"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="S3_distclust_files/figure-html/lk1mb-2.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="redo-clustering-with-alternative-distance-and-agglomeration-method">Redo clustering with alternative distance and agglomeration method<a class="anchor" aria-label="anchor" href="#redo-clustering-with-alternative-distance-and-agglomeration-method"></a>
</h4>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="va">cd</span>, method<span class="op">=</span><span class="st">"ward.D2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fl">.3</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="S3_distclust_files/figure-html/redo-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="silhouette-measure">Silhouette measure<a class="anchor" aria-label="anchor" href="#silhouette-measure"></a>
</h4>
<p>From ?silhouette with the cluster library:</p>
<pre><code>For each observation i, the _silhouette width_ s(i) is defined as follows:

     Put a(i) = average dissimilarity between i and all other points of
     the cluster to which i belongs (if i is the _only_ observation in
     its cluster, s(i) := 0 without further calculations).  For all
     _other_ clusters C, put d(i,C) = average dissimilarity of i to all
     observations of C.  The smallest of these d(i,C) is b(i) := \min_C
     d(i,C), and can be seen as the dissimilarity between i and its
     "neighbor" cluster, i.e., the nearest one to which it does _not_
     belong.  Finally,

                   s(i) := ( b(i) - a(i) ) / max( a(i), b(i) ).         
     
     'silhouette.default()' is now based on C code donated by Romain
     Francois (the R version being still available as
     'cluster:::silhouette.default.R').

     Observations with a large s(i) (almost 1) are very well clustered,
     a small s(i) (around 0) means that the observation lies between
     two clusters, and observations with a negative s(i) are probably
     placed in the wrong cluster.</code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ct1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html" class="external-link">cutree</a></span><span class="op">(</span><span class="va">fit2</span>, h<span class="op">=</span><span class="fl">.3</span><span class="op">)</span></span>
<span><span class="va">c2</span> <span class="op">=</span> <span class="va">cnv_log_R</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">c2</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">c2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">ct1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sil</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/cluster/man/silhouette.html" class="external-link">silhouette</a></span><span class="op">(</span><span class="va">ct1</span>, <span class="va">cd</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sil</span><span class="op">)</span></span></code></pre></div>
<p><img src="S3_distclust_files/figure-html/init-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h4>
<p><strong>1: install bioDist and vjcitn/csamaDist. Use the code:</strong></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">csamaDist</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">cnv_log_R</span><span class="op">)</span></span>
<span><span class="va">hc1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="va">cnv_log_R</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">hc2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="fu">bioDist</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bioDist/man/cor.dist.html" class="external-link">cor.dist</a></span><span class="op">(</span><span class="va">cnv_log_R</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span>,<span class="op">]</span><span class="op">)</span>, method<span class="op">=</span><span class="st">"ward.D2"</span><span class="op">)</span> </span>
<span><span class="va">opar</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>no.readonly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">hc1</span>, main<span class="op">=</span><span class="st">"Euc, complete"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">hc2</span>, main<span class="op">=</span><span class="st">"1-Cor, Ward's D2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span><span class="va">opar</span><span class="op">)</span></span></code></pre></div>
<p><strong>Comment on the qualitative differences between the clusterings</strong>.</p>
<p><strong>2: Here is how to produce a report on silhouette measurement for the second clustering.</strong></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">assn</span> <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/stats/cutree.html" class="external-link">cutree</a></span><span class="op">(</span><span class="va">hc2</span>, h<span class="op">=</span><span class="fl">.25</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/cluster/man/silhouette.html" class="external-link">silhouette</a></span><span class="op">(</span> <span class="va">assn</span>, <span class="fu">bioDist</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bioDist/man/cor.dist.html" class="external-link">cor.dist</a></span><span class="op">(</span><span class="va">cnv_log_R</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><strong>Produce a three-cluster partition from <code>hc1</code> and obtain the silhouette display.</strong></p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vince Carey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
