<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>S5_inference: FDR, linear models, and GLMs • CSHstats</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="../deps/MathJax-3.2.2/tex-chtml.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="S5_inference: FDR, linear models, and GLMs">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">CSHstats</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.34</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/S1_probability.html">S1_probability: Definitions and exercises</a></li>
    <li><a class="dropdown-item" href="../articles/S1b_hyper.html">S1a: A look at the hypergeometric distribution</a></li>
    <li><a class="dropdown-item" href="../articles/S2_EDA.html">S2_EDA: Exploratory data analysis and visualization</a></li>
    <li><a class="dropdown-item" href="../articles/S2a_PCAelem.html">S2a: Elements of PCA</a></li>
    <li><a class="dropdown-item" href="../articles/S3_distclust.html">S3: Distances, clustering</a></li>
    <li><a class="dropdown-item" href="../articles/S4_hypothesis_testing.html">S4_inference: Definitions and exercises</a></li>
    <li><a class="dropdown-item" href="../articles/S5_linear_models.html">S5_inference: FDR, linear models, and GLMs</a></li>
    <li><a class="dropdown-item" href="../articles/S5a_lmdiagnostics.html">S5a: Diagnostics of a differential gene expression exercise</a></li>
    <li><a class="dropdown-item" href="../articles/S5b_nonlin.html">S5b: hypothesis testing in nonlinear regression</a></li>
    <li><a class="dropdown-item" href="../articles/S6_hca_onto.html">S6 HCA, ontologies, shiny</a></li>
    <li><a class="dropdown-item" href="../articles/S7_cnn.html">S7 Neural net concepts and applications</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/vjcitn/CSHstats/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">


<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>S5_inference: FDR, linear models, and GLMs</h1>
                        <h4 data-toc-skip class="author">Vincent J. Carey, stvjc at channing.harvard.edu</h4>
            
            <h4 data-toc-skip class="date">January 09, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/CSHstats/blob/HEAD/vignettes/S5_linear_models.Rmd" class="external-link"><code>vignettes/S5_linear_models.Rmd</code></a></small>
      <div class="d-none name"><code>S5_linear_models.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="review-of-some-inference-concepts">Review of some inference concepts<a class="anchor" aria-label="anchor" href="#review-of-some-inference-concepts"></a>
</h2>
<div class="section level3">
<h3 id="error-rates-size-and-power">Error rates: size and power<a class="anchor" aria-label="anchor" href="#error-rates-size-and-power"></a>
</h3>
<ul>
<li>
<p>Null hypothesis testing framework: we typically aim to assess whether
an intervention affects a parameter value for a certain population</p>
<ul>
<li>does the average value of some quantity change after treatment?</li>
<li>does changing the composition of a deck of cards affect the
probability of seeing a given suit when drawing a card?</li>
</ul>
</li>
<li><p>Type I error: reject the null when it is actually true</p></li>
<li><p>Type II error: fail to reject null when it is actually false</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="p-value-a-condensation-of-the-data-and-assumptions">p-value: a condensation of the data and assumptions<a class="anchor" aria-label="anchor" href="#p-value-a-condensation-of-the-data-and-assumptions"></a>
</h3>
<p>The p-value for a test is the probability of observing the
statistic seen in the experiment <strong>or any more extreme
value of the statistic</strong> given that the null hypothesis is true.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/binom.test.html" class="external-link">binom.test</a></span><span class="op">(</span><span class="fl">27</span>, <span class="fl">100</span>, <span class="fl">1</span><span class="op">/</span><span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  Exact binomial test</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## data:  27 and 100</span></span>
<span><span class="co">## number of successes = 27, number of trials = 100, p-value = 0.6446</span></span>
<span><span class="co">## alternative hypothesis: true probability of success is not equal to 0.25</span></span>
<span><span class="co">## 95 percent confidence interval:</span></span>
<span><span class="co">##  0.1860664 0.3680163</span></span>
<span><span class="co">## sample estimates:</span></span>
<span><span class="co">## probability of success </span></span>
<span><span class="co">##                   0.27</span></span></code></pre>
<p>This two-sided p-value is obtained via</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">dbinom</a></span><span class="op">(</span><span class="fl">27</span><span class="op">:</span><span class="fl">100</span>, <span class="fl">100</span>, <span class="fl">.25</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">dbinom</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>,<span class="fl">100</span>, <span class="fl">.25</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.6446304</span></span></code></pre>
<p>because all the results of 27 or more hearts seen, or 22 or fewer
hearts seen are “as or more extreme” as what we have observed.</p>
<p>Because the p-value is a probability, its value reflects uncertainty
in our assessment of the relation of the data to null hypothesis.
Large p-values suggest that there is not much reason to use the
data to reject the null hypothesis; small p-values suggest that either
the null hypothesis is false or a very rare event has occurred.</p>
</div>
<div class="section level3">
<h3 id="confidence-interval-an-alternative-expression-of-uncertainty">Confidence interval: an alternative expression of uncertainty<a class="anchor" aria-label="anchor" href="#confidence-interval-an-alternative-expression-of-uncertainty"></a>
</h3>
<p>If we observed 30 hearts in 100 top-card draws, our test report would be:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/binom.test.html" class="external-link">binom.test</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">100</span>, <span class="fl">1</span><span class="op">/</span><span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  Exact binomial test</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## data:  30 and 100</span></span>
<span><span class="co">## number of successes = 30, number of trials = 100, p-value = 0.2491</span></span>
<span><span class="co">## alternative hypothesis: true probability of success is not equal to 0.25</span></span>
<span><span class="co">## 95 percent confidence interval:</span></span>
<span><span class="co">##  0.2124064 0.3998147</span></span>
<span><span class="co">## sample estimates:</span></span>
<span><span class="co">## probability of success </span></span>
<span><span class="co">##                    0.3</span></span></code></pre>
<p>The confidence interval is a random interval derived from the data
that has the property that it will include the true value of the
population parameter of interest with a specified probability.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">stats</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">27</span><span class="op">:</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">tests</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">stats</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/binom.test.html" class="external-link">binom.test</a></span><span class="op">(</span><span class="va">x</span>, <span class="fl">100</span>, <span class="fl">.25</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">low</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">tests</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">$</span><span class="va">conf.int</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">hi</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">tests</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">$</span><span class="va">conf.int</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">ests</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">tests</span>, <span class="st">"[["</span>, <span class="st">"statistic"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">stats</span>, <span class="va">ests</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">.6</span><span class="op">)</span>, main<span class="op">=</span><span class="st">"95% Confidence intervals"</span>, xlab<span class="op">=</span><span class="st">"# hearts seen in 100 draws"</span>,</span>
<span>  ylab<span class="op">=</span><span class="st">"estimated proportion of hearts"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/segments.html" class="external-link">segments</a></span><span class="op">(</span> <span class="va">stats</span>, <span class="va">low</span>, <span class="va">stats</span>, <span class="va">hi</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>h<span class="op">=</span><span class="fl">.25</span>, lty<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="S5_linear_models_files/figure-html/lkcis-1.png" width="672"></p>
<p>Exercise: For what value of the statistic would you reject the null hypothesis of
probability 1/4 for heart?</p>
<p>Exercise: Use 99% as the coverage probability and produce the display.</p>
</div>
</div>
<div class="section level2">
<h2 id="multiple-comparisons">Multiple comparisons<a class="anchor" aria-label="anchor" href="#multiple-comparisons"></a>
</h2>
<p>Everything we’ve seen thus far looks at <strong>a single test</strong> in various ways.</p>
<p>A hallmark of work in genomics is the necessity of performing many tests, because
of the large number of features and hypotheses in play.</p>
<p>Fact: When many tests are conducted on true null hypotheses, the distribution
of the collection of p-values thereby obtained is uniform on (0,1].</p>
<p>To illustrate this we will produce 10000 samples from N(100,1) and conduct
the t test with null mean 100, yielding 10000 p-values.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">many_x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">replicate</a></span><span class="op">(</span><span class="fl">10000</span>, <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">many_p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">many_x</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/t.test.html" class="external-link">t.test</a></span><span class="op">(</span><span class="va">x</span>, mu<span class="op">=</span><span class="fl">100</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/density.html" class="external-link">density</a></span><span class="op">(</span><span class="va">many_p</span>, from<span class="op">=</span><span class="fl">0</span>, to<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="S5_linear_models_files/figure-html/lknull-1.png" width="672"></p>
<p>This display shows a limitation of density estimation on a fixed interval –
a “boundary effect”, because data “up against” the boundary
get sparser as you get closer to the boundary. But it is consistent with
the property of a uniform distribution on [0,1]: the true density is 1.</p>
<div class="section level3">
<h3 id="bonferronis-correction-to-achieve-family-wise-error-rate-fwer-control">Bonferroni’s correction to achieve Family-Wise Error Rate (FWER) control<a class="anchor" aria-label="anchor" href="#bonferronis-correction-to-achieve-family-wise-error-rate-fwer-control"></a>
</h3>
<p>The concept of Type I error is very general, and could be deployed
to define operating characteristics for any kind of inference procedure.</p>
<p>This table is frequently seen in discussions of multiple testing, it is
from Holmes and Huber’s Modern Statistics for Modern Biology.</p>
<div class="figure">
<img src="MTESTTAB.jpg" alt=""><p class="caption">decisiontab</p>
</div>
<p>Suppose we have <span class="math inline">\(m\)</span> hypotheses to test and we wish the overall
probability of our family of tests to have Type I error rate <span class="math inline">\(\alpha\)</span>.
Using the table symbols, this is <span class="math inline">\(Pr(V&gt;0) &lt; \alpha\)</span>.</p>
<p>Bonferroni’s method is to declare significance only for
those tests with <span class="math inline">\(p\)</span>-value less than <span class="math inline">\(\alpha/m\)</span>.</p>
<p>This procedure (and related improvements
for FWER control) also implies a transformation of <span class="math inline">\(p\)</span>-values, that we will
see shortly.</p>
</div>
<div class="section level3">
<h3 id="false-discovery-rate">False discovery rate<a class="anchor" aria-label="anchor" href="#false-discovery-rate"></a>
</h3>
<p>Referring to the table above, the false discovery rate (FDR) is
the expected value of <span class="math inline">\(V/max(R,1)\)</span>. If <span class="math inline">\(R = 0\)</span>, <span class="math inline">\(V = 0\)</span> and FDR is zero.
Otherwise, we can think of FDR as a kind of budget: we will reject <span class="math inline">\(R\)</span>
hypotheses and our expection is that <span class="math inline">\(V\)</span> will be false. If we are
“following up” with confirmatory experiments on <span class="math inline">\(R\)</span> genes and can
afford “wasting” <span class="math inline">\(V\)</span> tests, we would accept an FDR of <span class="math inline">\(V/R\)</span>.</p>
<p>Of course these are probabilistic characteristics of procedures that
hold with stated probabilities when assumptions are satisfied.</p>
</div>
<div class="section level3">
<h3 id="adjusted-p-values">Adjusted p-values<a class="anchor" aria-label="anchor" href="#adjusted-p-values"></a>
</h3>
<p>Since we are so accustomed to the single-number summary <span class="math inline">\(p\)</span>, it is
typical to base rejection decisions on transformed <span class="math inline">\(p\)</span>-values. The
multtest package takes care of this.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">multtest</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>digits<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">tab</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/multtest/man/mt.rawp2adjp.html" class="external-link">mt.rawp2adjp</a></span><span class="op">(</span><span class="va">many_p</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tab</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           rawp Bonferroni  Holm Hochberg SidakSS SidakSD    BH BY   ABH</span></span>
<span><span class="co">##  [1,] 9.27e-05      0.927 0.927    0.927   0.604   0.604 0.645  1 0.645</span></span>
<span><span class="co">##  [2,] 1.81e-04      1.000 1.000    1.000   0.836   0.836 0.645  1 0.645</span></span>
<span><span class="co">##  [3,] 3.93e-04      1.000 1.000    1.000   0.980   0.980 0.645  1 0.645</span></span>
<span><span class="co">##  [4,] 4.35e-04      1.000 1.000    1.000   0.987   0.987 0.645  1 0.645</span></span>
<span><span class="co">##  [5,] 4.40e-04      1.000 1.000    1.000   0.988   0.988 0.645  1 0.645</span></span>
<span><span class="co">##  [6,] 4.43e-04      1.000 1.000    1.000   0.988   0.988 0.645  1 0.645</span></span>
<span><span class="co">##  [7,] 4.96e-04      1.000 1.000    1.000   0.993   0.993 0.645  1 0.645</span></span>
<span><span class="co">##  [8,] 6.28e-04      1.000 1.000    1.000   0.998   0.998 0.645  1 0.645</span></span>
<span><span class="co">##  [9,] 6.82e-04      1.000 1.000    1.000   0.999   0.999 0.645  1 0.645</span></span>
<span><span class="co">## [10,] 7.14e-04      1.000 1.000    1.000   0.999   0.999 0.645  1 0.645</span></span>
<span><span class="co">##       TSBH_0.05</span></span>
<span><span class="co">##  [1,]     0.645</span></span>
<span><span class="co">##  [2,]     0.645</span></span>
<span><span class="co">##  [3,]     0.645</span></span>
<span><span class="co">##  [4,]     0.645</span></span>
<span><span class="co">##  [5,]     0.645</span></span>
<span><span class="co">##  [6,]     0.645</span></span>
<span><span class="co">##  [7,]     0.645</span></span>
<span><span class="co">##  [8,]     0.645</span></span>
<span><span class="co">##  [9,]     0.645</span></span>
<span><span class="co">## [10,]     0.645</span></span></code></pre>
<p>By default, <code>mt.rawp2adp</code> produces a number of adjustments – we are concerned
at present with the columns rawp (the p-values sorted from lowest on), Bonferroni
(which gives the transformation <span class="math inline">\(p\)</span> to min(1, mp)$ for <span class="math inline">\(m\)</span> tests),
and BH (the Benjamini-Hochberg FDR transformation). The implication
of this table is that with FWER control at 0.05, our data
would not support rejection of any of our
hypotheses (all of which assert that the mean of a sample of 10 random
normals is equal to 100). Likewise with FDR control at any value up to
0.77. However, we see that had we accepted an FDR control at 0.8,
we could reject 8 hypotheses.</p>
<p>Exercise: replace the first p-value in <code>many_p</code> with <code>1e-6</code> and obtain the
associated adjusted p-values.</p>
</div>
<div class="section level3">
<h3 id="adjustments-via-weighted-fdr">Adjustments via weighted FDR<a class="anchor" aria-label="anchor" href="#adjustments-via-weighted-fdr"></a>
</h3>
<p>A <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10276986/" class="external-link">useful review</a> is
available. This section addresses only one of the methods in the review, which
is found to have good performance, though in some contexts the authors’ alternatives
appear superior.</p>
<div class="section level4">
<h4 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h4>
<p>We need the IHW package and will do some visualization as well.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/CSHstats" class="external-link">CSHstats</a></span><span class="op">)</span></span>
<span> <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span> <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com" class="external-link">plotly</a></span><span class="op">)</span></span>
<span> <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">IHW</span><span class="op">)</span></span>
<span> <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">multtest</span><span class="op">)</span></span></code></pre></div>
<p>Here’s the key resource with p-values that we want to interpret.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">gtex_exc_chr20_b38</span><span class="op">)</span></span>
<span><span class="va">gtdat</span> <span class="op">=</span> <span class="va">gtex_exc_chr20_b38</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gtdat</span>,<span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   chr pos_b38       marker         rsid effectallele otherallele    eaf     p</span></span>
<span><span class="co">## 1  20   61066 20:61066:T:C rs1031667049            C           T 0.0285 0.774</span></span>
<span><span class="co">## 2  20   61083 20:61083:C:T  rs956068117            T           C 0.1010 0.767</span></span>
<span><span class="co">## 3  20   61098 20:61098:C:T  rs933208784            T           C 0.0155 0.117</span></span>
<span><span class="co">##      beta     se              pheno mode       tag imputersq   n     Gene</span></span>
<span><span class="co">## 1  0.0226 0.0784 ENSG00000196476.11 eqtl GTEx-eqtl        NA 515 C20orf96</span></span>
<span><span class="co">## 2  0.0123 0.0413 ENSG00000196476.11 eqtl GTEx-eqtl        NA 515 C20orf96</span></span>
<span><span class="co">## 3 -0.1615 0.1029 ENSG00000196476.11 eqtl GTEx-eqtl        NA 515 C20orf96</span></span>
<span><span class="co">##   seqnames  start    end width strand         gene_id gene_name   gene_biotype</span></span>
<span><span class="co">## 1       20 270863 290778 19916      - ENSG00000196476  C20orf96 protein_coding</span></span>
<span><span class="co">## 2       20 270863 290778 19916      - ENSG00000196476  C20orf96 protein_coding</span></span>
<span><span class="co">## 3       20 270863 290778 19916      - ENSG00000196476  C20orf96 protein_coding</span></span>
<span><span class="co">##   seq_coord_system   symbol entrezid</span></span>
<span><span class="co">## 1       chromosome C20orf96   140680</span></span>
<span><span class="co">## 2       chromosome C20orf96   140680</span></span>
<span><span class="co">## 3       chromosome C20orf96   140680</span></span></code></pre>
<p>What we have here are variant-expression association tests conducted
in the <a href="https://gtexportal.org" class="external-link">GTEx project</a>. Thus this is a small
excerpt from a collection of “GWAS”, in which the outcome is tissue-specific
expression of a gene, and predictors are formed by SNP genotypes and
adjustments carried out in the GTEx analysis pipeline.</p>
<p>Here’s a composite manhattan plot:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pl1</span> <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gtdat</span><span class="op">[</span><span class="va">gtdat</span><span class="op">$</span><span class="va">p</span><span class="op">&lt;</span><span class="fl">.1</span>,<span class="op">]</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">pos_b38</span>, y<span class="op">=</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span>, colour<span class="op">=</span><span class="va">Gene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#ggplotly(pl1)</span></span>
<span><span class="va">pl1</span></span></code></pre></div>
<p><img src="S5_linear_models_files/figure-html/lkmanh-1.png" width="672"></p>
</div>
<div class="section level4">
<h4 id="demonstration-1-use-allele-frequency-to-form-strata-for-fdr-weighting">Demonstration 1: Use allele frequency to form strata for FDR weighting<a class="anchor" aria-label="anchor" href="#demonstration-1-use-allele-frequency-to-form-strata-for-fdr-weighting"></a>
</h4>
<p>The <code>eaf</code> field is the frequency with which the “effect allele” was
reported by SNP-Nexus.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">covariate1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">gtdat</span><span class="op">$</span><span class="va">eaf</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">.01</span>,<span class="fl">.05</span>,<span class="fl">.1</span>,<span class="fl">.3</span>,<span class="fl">.6</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">covariate1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## covariate1</span></span>
<span><span class="co">##    (0,0.01] (0.01,0.05]  (0.05,0.1]   (0.1,0.3]   (0.3,0.6] </span></span>
<span><span class="co">##        3865       33697       13860       27840       20738</span></span></code></pre>
<p>We run the IHW algorithm using the variant frequency strata, with
a conjecture that very rare or very common variants may be downweighted
in a way that is advantageous to discovery.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IHW/man/ihw.default.html" class="external-link">ihw</a></span><span class="op">(</span><span class="va">gtdat</span><span class="op">$</span><span class="va">p</span>, <span class="va">covariate1</span>, alpha<span class="op">=</span><span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">basic_fdr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/multtest/man/mt.rawp2adjp.html" class="external-link">mt.rawp2adjp</a></span><span class="op">(</span><span class="va">gtdat</span><span class="op">$</span><span class="va">p</span><span class="op">)</span></span></code></pre></div>
<p>The weighting procedure involves random resampling from the data,
to reduce risk of overfitting, and to allow assessment of
consistency across multiple “folds” or random subsets.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">demo1</span><span class="op">)</span></span></code></pre></div>
<p><img src="S5_linear_models_files/figure-html/lkwts1-1.png" width="672"></p>
<p>Compare the weighted, unweighted, and conventional
rejection numbers:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">newp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IHW/man/ihwResult-class.html" class="external-link">adj_pvalues</a></span><span class="op">(</span><span class="va">demo1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">newp</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 276</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">basic_fdr</span><span class="op">$</span><span class="va">adjp</span><span class="op">[</span>,<span class="st">"BH"</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 257</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">gtdat</span><span class="op">$</span><span class="va">p</span> <span class="op">&lt;</span> <span class="fl">5e-8</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 68</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="demonstration-2-use-distance-to-gene-to-weight">Demonstration 2: Use distance to gene to weight<a class="anchor" aria-label="anchor" href="#demonstration-2-use-distance-to-gene-to-weight"></a>
</h4>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dist2gene</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">gtdat</span><span class="op">$</span><span class="va">pos_b38</span><span class="op">-</span><span class="va">gtdat</span><span class="op">$</span><span class="va">start</span><span class="op">)</span></span>
<span><span class="va">covariate2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">dist2gene</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">ok</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">covariate2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">demo2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IHW/man/ihw.default.html" class="external-link">ihw</a></span><span class="op">(</span><span class="va">gtdat</span><span class="op">$</span><span class="va">p</span><span class="op">[</span><span class="va">ok</span><span class="op">]</span>, <span class="va">covariate2</span><span class="op">[</span><span class="va">ok</span><span class="op">]</span>, alpha<span class="op">=</span><span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">demo2</span><span class="op">)</span></span></code></pre></div>
<p><img src="S5_linear_models_files/figure-html/lkposi-1.png" width="672"></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">newp2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IHW/man/ihwResult-class.html" class="external-link">adj_pvalues</a></span><span class="op">(</span><span class="va">demo2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">newp2</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 379</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="optional-considerations-on-bayesian-multiple-testing">Optional: Considerations on Bayesian multiple testing<a class="anchor" aria-label="anchor" href="#optional-considerations-on-bayesian-multiple-testing"></a>
</h3>
<p>Various comments and references in this <a href="https://stats.stackexchange.com/questions/203378/why-dont-bayesian-methods-require-multiple-testing-corrections" class="external-link">stackexchange</a> address the relationship
between inferential framework and approach to multiple testing. A <a href="http://www.stat.columbia.edu/~gelman/research/unpublished/p_hacking.pdf" class="external-link">paper
by Gelman and Loken</a>
surveys the concept in an illuminating way.</p>
<p>A paper in the <a href="https://doi.org/10.1007/s10654-019-00517-2" class="external-link">European Journal of Epidemiology</a>
presents considerations on multiple comparisons from a Bayesian perspective, arguing
that the issue is “surrounded by confusion and controversy”.</p>
<p>Here is a simple approach to visualizing 20 tests, assumed independent,
in a frequenist framework, focusing on 95% confidence intervals.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># from https://link.springer.com/content/pdf/10.1007/s10654-019-00517-2.pdf</span></span>
<span><span class="co"># Sjolander and vanSteenlandt 2019 Eur J Epi</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://gforge.se/packages/" class="external-link">forestplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bayesm</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="va">J</span> <span class="op">=</span> <span class="fl">20</span></span>
<span><span class="va">beta</span> <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="va">Y</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">n</span><span class="op">*</span><span class="va">J</span>, mean<span class="op">=</span><span class="va">beta</span><span class="op">)</span>, nrow<span class="op">=</span><span class="va">n</span>, ncol<span class="op">=</span><span class="va">J</span><span class="op">)</span></span>
<span><span class="va">est</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">Y</span>, MARGIN<span class="op">=</span><span class="fl">2</span>, FUN<span class="op">=</span><span class="va">sd</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span></span>
<span><span class="va">q</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/TDist.html" class="external-link">qt</a></span><span class="op">(</span>p<span class="op">=</span><span class="fl">0.975</span>, df<span class="op">=</span><span class="va">n</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">ci.l</span> <span class="op">=</span> <span class="va">est</span><span class="op">-</span><span class="va">q</span><span class="op">*</span><span class="va">se</span></span>
<span><span class="va">ci.u</span> <span class="op">=</span> <span class="va">est</span><span class="op">+</span><span class="va">q</span><span class="op">*</span><span class="va">se</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/forestplot/man/forestplot.html" class="external-link">forestplot</a></span><span class="op">(</span></span>
<span> labeltext<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">""</span>, <span class="va">J</span><span class="op">)</span>,</span>
<span> mean<span class="op">=</span><span class="va">est</span>, lower<span class="op">=</span><span class="va">ci.l</span>,</span>
<span> upper<span class="op">=</span><span class="va">ci.u</span>, ci.vertices<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span> boxsize<span class="op">=</span><span class="fl">0.3</span>, txt_gp<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/forestplot/man/fpTxtGp.html" class="external-link">fpTxtGp</a></span><span class="op">(</span>cex<span class="op">=</span><span class="fl">2</span><span class="op">)</span>,</span>
<span> xticks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">.5</span><span class="op">)</span>, title<span class="op">=</span><span class="st">"Frequentist 95% confidence intervals"</span><span class="op">)</span></span></code></pre></div>
<p><img src="S5_linear_models_files/figure-html/lkfor1-1.png" width="672"></p>
<p>One hypothesis is rejected.</p>
<p>The authors also use the <code>bayesm</code> package to produce
Bayesian credible intervals <strong>under the assumption that all tests are independent</strong>.</p>
<p><img src="S5_linear_models_files/figure-html/lkfp1-1.png" width="672"></p>
<p>Finally, allowing high correlatedness among test statistics,
the Bayesian approach is found to avoid any false rejections.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/forestplot/man/forestplot.html" class="external-link">forestplot</a></span><span class="op">(</span></span>
<span> labeltext<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">""</span>, <span class="va">J</span><span class="op">)</span>,</span>
<span> mean<span class="op">=</span><span class="va">est</span>, lower<span class="op">=</span><span class="va">ci.l</span>,</span>
<span> upper<span class="op">=</span><span class="va">ci.u</span>, ci.vertices<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span> boxsize<span class="op">=</span><span class="fl">0.3</span>, txt_gp<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/forestplot/man/fpTxtGp.html" class="external-link">fpTxtGp</a></span><span class="op">(</span>cex<span class="op">=</span><span class="fl">2</span><span class="op">)</span>,</span>
<span> xticks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.5</span>,<span class="fl">1.5</span>,<span class="fl">.5</span><span class="op">)</span>, title<span class="op">=</span><span class="st">"Bayesian, assume correlated tests"</span><span class="op">)</span></span></code></pre></div>
<p><img src="S5_linear_models_files/figure-html/lkatnew-1.png" width="672"></p>
<p>The Bayesian computations involve Monte Carlo Markov Chain simulation.
Background can be obtained from <a href="https://xcelab.net/rm/statistical-rethinking/" class="external-link">Statistical Rethinking</a>
by R. McElreath.</p>
</div>
</div>
<div class="section level2">
<h2 id="linear-models">Linear models<a class="anchor" aria-label="anchor" href="#linear-models"></a>
</h2>
<div class="section level3">
<h3 id="data-setup">Data setup<a class="anchor" aria-label="anchor" href="#data-setup"></a>
</h3>
<p>Let’s return to the TCGA ACC FOS and EGR1 data that we used to study correlation.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">fos_ex</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">egr1_ex</span><span class="op">)</span></span>
<span><span class="va">fedf</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>FOS<span class="op">=</span><span class="va">fos_ex</span>, EGR1<span class="op">=</span><span class="va">egr1_ex</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">fedf</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">EGR1</span>, y<span class="op">=</span><span class="va">FOS</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="S5_linear_models_files/figure-html/lkfos-1.png" width="672"></p>
<p>We found that log transformation was useful for ameliorating
the overplotting near the origin:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">fedf</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">EGR1</span>, y<span class="op">=</span><span class="va">FOS</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="S5_linear_models_files/figure-html/dopl2-1.png" width="672"></p>
<p>The ACC data are annotated in a MultiAssayExperiment, and a mutation
summary called OncoSign is available.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://waldronlab.io/MultiAssayExperiment/" class="external-link">MultiAssayExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/CSHstats" class="external-link">CSHstats</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">accex</span><span class="op">)</span></span>
<span><span class="va">accex</span></span></code></pre></div>
<pre><code><span><span class="co">## A MultiAssayExperiment object of 1 listed</span></span>
<span><span class="co">##  experiment with a user-defined name and respective class.</span></span>
<span><span class="co">##  Containing an ExperimentList class object of length 1:</span></span>
<span><span class="co">##  [1] ACC_RNASeq2GeneNorm-20160128: SummarizedExperiment with 20501 rows and 79 columns</span></span>
<span><span class="co">## Functionality:</span></span>
<span><span class="co">##  experiments() - obtain the ExperimentList instance</span></span>
<span><span class="co">##  colData() - the primary/phenotype DataFrame</span></span>
<span><span class="co">##  sampleMap() - the sample coordination DataFrame</span></span>
<span><span class="co">##  `$`, `[`, `[[` - extract colData columns, subset, or experiment</span></span>
<span><span class="co">##  *Format() - convert into a long or wide DataFrame</span></span>
<span><span class="co">##  assays() - convert ExperimentList to a SimpleList of matrices</span></span>
<span><span class="co">##  exportClass() - save data to flat files</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">accex</span><span class="op">$</span><span class="va">OncoSign</span>, exclude<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##          CN1          CN2       CTNNB1   TERT/ZNRF3     TP53/NF1 Unclassified </span></span>
<span><span class="co">##            4           13            5           17           21           14 </span></span>
<span><span class="co">##         &lt;NA&gt; </span></span>
<span><span class="co">##            5</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="re-expressing-the-two-sample-test">Re-expressing the two-sample test<a class="anchor" aria-label="anchor" href="#re-expressing-the-two-sample-test"></a>
</h3>
<p>We will focus on two OncoSign classes, asking whether average log EGR1
expression is different between them.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hasonc</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">accex</span><span class="op">$</span><span class="va">OncoSign</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ndf</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>logEGR1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">egr1_ex</span><span class="op">[</span><span class="va">hasonc</span><span class="op">]</span><span class="op">)</span>, oncosign<span class="op">=</span><span class="va">accex</span><span class="op">$</span><span class="va">OncoSign</span><span class="op">[</span><span class="va">hasonc</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">tpvtert</span> <span class="op">=</span> <span class="va">ndf</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">oncosign</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TP53/NF1"</span>, <span class="st">"TERT/ZNRF3"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">tpvtert</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="va">logEGR1</span>, x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">oncosign</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="S5_linear_models_files/figure-html/reform-1.png" width="672"></p>
<div class="section level4">
<h4 id="there-are-two-t-tests">There are two t tests<a class="anchor" aria-label="anchor" href="#there-are-two-t-tests"></a>
</h4>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>digits<span class="op">=</span><span class="fl">6</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html" class="external-link">t.test</a></span><span class="op">(</span><span class="va">logEGR1</span><span class="op">~</span><span class="va">oncosign</span>, data<span class="op">=</span><span class="va">tpvtert</span>, var.equal<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  Two Sample t-test</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## data:  logEGR1 by oncosign</span></span>
<span><span class="co">## t = 1.038, df = 36, p-value = 0.306</span></span>
<span><span class="co">## alternative hypothesis: true difference in means between group TERT/ZNRF3 and group TP53/NF1 is not equal to 0</span></span>
<span><span class="co">## 95 percent confidence interval:</span></span>
<span><span class="co">##  -0.374199  1.158895</span></span>
<span><span class="co">## sample estimates:</span></span>
<span><span class="co">## mean in group TERT/ZNRF3   mean in group TP53/NF1 </span></span>
<span><span class="co">##                  7.35317                  6.96083</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html" class="external-link">t.test</a></span><span class="op">(</span><span class="va">logEGR1</span><span class="op">~</span><span class="va">oncosign</span>, data<span class="op">=</span><span class="va">tpvtert</span>, var.equal<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="co"># more general</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  Welch Two Sample t-test</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## data:  logEGR1 by oncosign</span></span>
<span><span class="co">## t = 1.046, df = 35.19, p-value = 0.303</span></span>
<span><span class="co">## alternative hypothesis: true difference in means between group TERT/ZNRF3 and group TP53/NF1 is not equal to 0</span></span>
<span><span class="co">## 95 percent confidence interval:</span></span>
<span><span class="co">##  -0.369343  1.154039</span></span>
<span><span class="co">## sample estimates:</span></span>
<span><span class="co">## mean in group TERT/ZNRF3   mean in group TP53/NF1 </span></span>
<span><span class="co">##                  7.35317                  6.96083</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="the-linear-model-has-one-of-these-as-a-special-case">The linear model has one of these as a special case<a class="anchor" aria-label="anchor" href="#the-linear-model-has-one-of-these-as-a-special-case"></a>
</h4>
<p>The general form of linear model in statistics has
“dependent variable” <span class="math inline">\(y\)</span>
<span class="math display">\[
y = \alpha + x' \beta + e
\]</span>
where <span class="math inline">\(x\)</span> is a p-vector of “covariates” – measurements that
are believed to affect the mean value of <span class="math inline">\(y\)</span>, also
known as “independent variables”, and <span class="math inline">\(e\)</span> is a zero-mean
disturbance or residual with a fixed dispersion. The parameters
of interest are <span class="math inline">\(\alpha\)</span> (often called intercept) and <span class="math inline">\(\beta\)</span>.
<strong>Parameter interpretation in terms of units of the response
relative to differences between samples with different values of <span class="math inline">\(x\)</span>
should always be explicit.</strong></p>
<div class="section level5">
<h5 id="a-dummy-variable-design-matrix">A dummy variable design matrix<a class="anchor" aria-label="anchor" href="#a-dummy-variable-design-matrix"></a>
</h5>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mm</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">oncosign</span>, data<span class="op">=</span><span class="va">tpvtert</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mm</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   (Intercept) oncosignTP53/NF1</span></span>
<span><span class="co">## 1           1                1</span></span>
<span><span class="co">## 2           1                1</span></span>
<span><span class="co">## 3           1                0</span></span>
<span><span class="co">## 4           1                1</span></span>
<span><span class="co">## 5           1                0</span></span>
<span><span class="co">## 6           1                1</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">mm</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    (Intercept) oncosignTP53/NF1</span></span>
<span><span class="co">## 33           1                0</span></span>
<span><span class="co">## 34           1                1</span></span>
<span><span class="co">## 35           1                0</span></span>
<span><span class="co">## 36           1                0</span></span>
<span><span class="co">## 37           1                1</span></span>
<span><span class="co">## 38           1                1</span></span></code></pre>
</div>
<div class="section level5">
<h5 id="recovering-the-t-test-with-lm">Recovering the t test with <code>lm</code><a class="anchor" aria-label="anchor" href="#recovering-the-t-test-with-lm"></a>
</h5>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tpvtert</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="va">logEGR1</span><span class="op">~</span><span class="va">oncosign</span>, data<span class="op">=</span><span class="va">_</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## lm(formula = logEGR1 ~ oncosign, data = tpvtert)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Residuals:</span></span>
<span><span class="co">##    Min     1Q Median     3Q    Max </span></span>
<span><span class="co">## -2.287 -0.721  0.167  0.625  2.549 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients:</span></span>
<span><span class="co">##                  Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span><span class="co">## (Intercept)         7.353      0.281   26.17   &lt;2e-16 ***</span></span>
<span><span class="co">## oncosignTP53/NF1   -0.392      0.378   -1.04     0.31    </span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Residual standard error: 1.16 on 36 degrees of freedom</span></span>
<span><span class="co">## Multiple R-squared:  0.0291, Adjusted R-squared:  0.00209 </span></span>
<span><span class="co">## F-statistic: 1.08 on 1 and 36 DF,  p-value: 0.306</span></span></code></pre>
<p>Exercise: Which t test is recovered in this analysis?</p>
<p>Exercise: Conduct the associated Wilcoxon test.</p>
<p>Exercise: Conduct the Gaussian rank test mentioned in the Gelman blog “Don’t do the Wilcoxon”.</p>
<p>Extra goody: To produce the coefficients “by hand”:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/solve.html" class="external-link">solve</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">mm</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">mm</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">mm</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/matmult.html" class="external-link">%*%</a></span><span class="va">tpvtert</span><span class="op">$</span><span class="va">logEGR1</span></span></code></pre></div>
<pre><code><span><span class="co">##                       [,1]</span></span>
<span><span class="co">## (Intercept)       7.353174</span></span>
<span><span class="co">## oncosignTP53/NF1 -0.392348</span></span></code></pre>
<p>programs <span class="math inline">\((X'X)^{-1}X'y\)</span></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="analysis-of-variance-f-test">Analysis of variance: F test<a class="anchor" aria-label="anchor" href="#analysis-of-variance-f-test"></a>
</h3>
<p>The analysis of variance is focused on comparison of multiple groups. The
EGR1-OncoSign relationship is a reasonable example where this could be used.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">ndf</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y<span class="op">=</span><span class="va">logEGR1</span>, x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">oncosign</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="S5_linear_models_files/figure-html/lkggaov-1.png" width="672"></p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nda</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/aov.html" class="external-link">aov</a></span><span class="op">(</span><span class="va">logEGR1</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">oncosign</span><span class="op">)</span>, data<span class="op">=</span><span class="va">ndf</span><span class="op">)</span></span>
<span><span class="va">nda</span></span></code></pre></div>
<pre><code><span><span class="co">## Call:</span></span>
<span><span class="co">##    aov(formula = logEGR1 ~ factor(oncosign), data = ndf)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Terms:</span></span>
<span><span class="co">##                 factor(oncosign) Residuals</span></span>
<span><span class="co">## Sum of Squares           11.3084   92.1642</span></span>
<span><span class="co">## Deg. of Freedom                5        68</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Residual standard error: 1.1642</span></span>
<span><span class="co">## Estimated effects may be unbalanced</span></span></code></pre>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">nda</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                  Df Sum Sq Mean Sq F value Pr(&gt;F)</span></span>
<span><span class="co">## factor(oncosign)  5   11.3    2.26    1.67   0.15</span></span>
<span><span class="co">## Residuals        68   92.2    1.35</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="linear-regression-parameter-estimation-goodness-of-fit">Linear regression, parameter estimation, goodness of fit<a class="anchor" aria-label="anchor" href="#linear-regression-parameter-estimation-goodness-of-fit"></a>
</h3>
<p>More common applications of linear modeling take advantage
of continuity of a predictor. For a simple linear regression (one continuous
covariate), the interpretation of <span class="math inline">\(\beta\)</span> is: the difference in mean of <span class="math inline">\(y\)</span>
associated with a unit difference in <span class="math inline">\(x\)</span>.</p>
<p>Regarding log EGR1 as a predictor of log FOS, we have</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">fos_ex</span><span class="op">)</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">egr1_ex</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">m1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## lm(formula = log(fos_ex) ~ log(egr1_ex))</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Residuals:</span></span>
<span><span class="co">##    Min     1Q Median     3Q    Max </span></span>
<span><span class="co">## -1.545 -0.482  0.198  0.476  1.397 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients:</span></span>
<span><span class="co">##              Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span><span class="co">## (Intercept)    2.2980     0.4975    4.62  1.5e-05 ***</span></span>
<span><span class="co">## log(egr1_ex)   0.6895     0.0692    9.96  1.8e-15 ***</span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Residual standard error: 0.727 on 77 degrees of freedom</span></span>
<span><span class="co">## Multiple R-squared:  0.563,  Adjusted R-squared:  0.557 </span></span>
<span><span class="co">## F-statistic: 99.1 on 1 and 77 DF,  p-value: 1.75e-15</span></span></code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">m1</span><span class="op">)</span></span></code></pre></div>
<p><img src="S5_linear_models_files/figure-html/doreg-1.png" width="672"><img src="S5_linear_models_files/figure-html/doreg-2.png" width="672"><img src="S5_linear_models_files/figure-html/doreg-3.png" width="672"><img src="S5_linear_models_files/figure-html/doreg-4.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="confounding">Confounding<a class="anchor" aria-label="anchor" href="#confounding"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="glms-binary-counted-non-gaussian-responses">GLMs: binary, counted, non-Gaussian responses<a class="anchor" aria-label="anchor" href="#glms-binary-counted-non-gaussian-responses"></a>
</h2>
<p>The family of generalized linear models (GLM) embraces a number
of approaches to statistical inference on associations between
non-Gaussian responses and general configurations of covariates.</p>
<div class="section level3">
<h3 id="prologue-fishers-exact-test-for-2-x-2-table">Prologue: Fisher’s exact test for 2 x 2 table<a class="anchor" aria-label="anchor" href="#prologue-fishers-exact-test-for-2-x-2-table"></a>
</h3>
<p>We’ll work with FOS and EGR1 in an artificial way. We’ll
suppose that we have criteria for “activation”:</p>
<ul>
<li>FOS is activated if our RNA-seq measurement in ACC exceeds 1000</li>
<li>EGR1 is activated if our RNA-seq measurement in ACC exceeds 2000</li>
</ul>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/CSHstats" class="external-link">CSHstats</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">fos_ex</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">egr1_ex</span><span class="op">)</span></span>
<span><span class="va">fos_act</span> <span class="op">=</span> <span class="fl">1</span><span class="op">*</span><span class="op">(</span><span class="va">fos_ex</span> <span class="op">&gt;</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">egr1_act</span> <span class="op">=</span> <span class="fl">1</span><span class="op">*</span><span class="op">(</span><span class="va">egr1_ex</span> <span class="op">&gt;</span> <span class="fl">2000</span><span class="op">)</span></span></code></pre></div>
<p>We form a 2 x 2 cross classification of tumors using these definitions:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">acttab</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span>fos_act<span class="op">=</span><span class="va">fos_act</span>, egr1_act<span class="op">=</span><span class="va">egr1_act</span><span class="op">)</span></span>
<span><span class="va">acttab</span></span></code></pre></div>
<pre><code><span><span class="co">##        egr1_act</span></span>
<span><span class="co">## fos_act  0  1</span></span>
<span><span class="co">##       0 26  5</span></span>
<span><span class="co">##       1 25 23</span></span></code></pre>
<p>A basic question is whether the two events “EGR1 activated” and “FOS activated”
are statistically associated in ACC tumors.</p>
<p>Fisher’s exact test evaluates the hypergeometric distribution of the
(1,1) cell of the table, given the marginal totals.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/fisher.test.html" class="external-link">fisher.test</a></span><span class="op">(</span><span class="va">acttab</span>, alternative<span class="op">=</span><span class="st">"greater"</span><span class="op">)</span> <span class="co"># one-sided</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##  Fisher's Exact Test for Count Data</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## data:  acttab</span></span>
<span><span class="co">## p-value = 0.0034</span></span>
<span><span class="co">## alternative hypothesis: true odds ratio is greater than 1</span></span>
<span><span class="co">## 95 percent confidence interval:</span></span>
<span><span class="co">##  1.68701     Inf</span></span>
<span><span class="co">## sample estimates:</span></span>
<span><span class="co">## odds ratio </span></span>
<span><span class="co">##    4.69113</span></span></code></pre>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Hypergeometric.html" class="external-link">dhyper</a></span><span class="op">(</span><span class="fl">26</span><span class="op">:</span><span class="fl">31</span>, <span class="fl">51</span>, <span class="fl">28</span>, <span class="fl">31</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.00339765</span></span></code></pre>
<p>The odds of an event that occurs with probability <span class="math inline">\(p\)</span> is given by <span class="math inline">\(p/(1-p)\)</span>.
Abbreviate to <span class="math inline">\(odds(p)\)</span></p>
<p>The odds ratio for two events with probabilities <span class="math inline">\(p_1\)</span> and <span class="math inline">\(p_2\)</span>
is <span class="math inline">\(odds(p_1)/odds(p_2)\)</span>. With a 2x2 table this can be estimated with
the cross ratio “ad/bc”. But the odds ratio estimated in <code>fisher.test</code>
is different, a conditional maximum likelihood estimate. The difference
is seldom meaningful.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">odds</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span><span class="va">p</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>; <span class="va">p</span><span class="op">/</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">p</span><span class="op">)</span><span class="op">}</span></span>
<span><span class="fu">odds</span><span class="op">(</span><span class="va">fos_act</span><span class="op">[</span><span class="va">egr1_act</span><span class="op">==</span><span class="cn">TRUE</span><span class="op">]</span><span class="op">)</span><span class="op">/</span><span class="fu">odds</span><span class="op">(</span><span class="va">fos_act</span><span class="op">[</span><span class="va">egr1_act</span><span class="op">==</span><span class="cn">FALSE</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 4.784</span></span></code></pre>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">acttab</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">]</span><span class="op">*</span><span class="va">acttab</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">]</span><span class="op">/</span><span class="op">(</span><span class="va">acttab</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">]</span><span class="op">*</span><span class="va">acttab</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 4.784</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="binary-response-logistic-regression">Binary response: logistic regression<a class="anchor" aria-label="anchor" href="#binary-response-logistic-regression"></a>
</h3>
<p>We model the conditional probability that FOS is activated,
depending on whether or note EGR1 is activated, using the
definitions given above.
The model has the form</p>
<p><span class="math display">\[
\mbox{logit} Pr(FOS &gt; 1000|EGR1&gt;2000) = \alpha + \beta x
\]</span></p>
<p>Here logit(p) = log(p/(1-p)) is called a “link function”. It
transforms the outcome of interest (a probability) from the
interval (0,1) to the whole real line. The interpretation of
<span class="math inline">\(\beta\)</span> is: “log odds ratio”. If we exponentiate beta, we
obtain the effect of a change in x of one unit on the odds
of the response event.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></span><span class="op">(</span><span class="va">fos_act</span> <span class="op">~</span> <span class="va">egr1_act</span>, family<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/family.html" class="external-link">binomial</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">g1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## glm(formula = fos_act ~ egr1_act, family = binomial())</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients:</span></span>
<span><span class="co">##             Estimate Std. Error z value Pr(&gt;|z|)   </span></span>
<span><span class="co">## (Intercept)  -0.0392     0.2801   -0.14   0.8886   </span></span>
<span><span class="co">## egr1_act      1.5653     0.5673    2.76   0.0058 **</span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## (Dispersion parameter for binomial family taken to be 1)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Null deviance: 105.830  on 78  degrees of freedom</span></span>
<span><span class="co">## Residual deviance:  96.958  on 77  degrees of freedom</span></span>
<span><span class="co">## AIC: 101</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Fisher Scoring iterations: 3</span></span></code></pre>
<p>Parameter interpretation. The intercept parameter has value
-0.0392, which is the logit of the probability
that FOS is activated, given that EGR1 is NOT activated.</p>
<p>The “beta” parameter is the log odds ratio (comparing probability
of FOS activation between the EGR1 conditions – not activated and
activated).</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">4.784</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1.56528</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="multiple-logistic-regression">Multiple logistic regression<a class="anchor" aria-label="anchor" href="#multiple-logistic-regression"></a>
</h3>
<p>Fisher’s exact test and simple logistic regression are
seen to be consistent with one another in this example.
The advantage of the logistic regression framework is that
more potential sources of variation in probability of response
can be incorporated.</p>
<p>We have data on YY1 expression in ACC and will declare it activated
when expressed at a level greater than 1500.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">yy1_ex</span><span class="op">)</span></span>
<span><span class="va">yy1_act</span> <span class="op">=</span> <span class="va">yy1_ex</span> <span class="op">&gt;</span> <span class="fl">1500</span></span>
<span><span class="va">g2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm</a></span><span class="op">(</span><span class="va">fos_act</span> <span class="op">~</span> <span class="va">egr1_act</span><span class="op">+</span><span class="va">yy1_act</span>, family<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/family.html" class="external-link">binomial</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">g2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## glm(formula = fos_act ~ egr1_act + yy1_act, family = binomial())</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients:</span></span>
<span><span class="co">##             Estimate Std. Error z value Pr(&gt;|z|)   </span></span>
<span><span class="co">## (Intercept)   0.0922     0.3175    0.29   0.7716   </span></span>
<span><span class="co">## egr1_act      1.5690     0.5702    2.75   0.0059 **</span></span>
<span><span class="co">## yy1_actTRUE  -0.4833     0.5472   -0.88   0.3771   </span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## (Dispersion parameter for binomial family taken to be 1)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Null deviance: 105.830  on 78  degrees of freedom</span></span>
<span><span class="co">## Residual deviance:  96.177  on 76  degrees of freedom</span></span>
<span><span class="co">## AIC: 102.2</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Fisher Scoring iterations: 3</span></span></code></pre>
<p>The inclusion of information on YY1 does not seem to be
informative on FOS activation.</p>
</div>
<div class="section level3">
<h3 id="general-case-link-and-variance-functions">General case: link and variance functions<a class="anchor" aria-label="anchor" href="#general-case-link-and-variance-functions"></a>
</h3>
<p>The family of generalized linear models can accommodate many
forms of variation in a response of interest, to which are
associated a number of “predictor” variables.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">glmtab</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">dists</th>
<th align="left">links</th>
<th align="left">varfuns</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">gaussian</td>
<td align="left">identity</td>
<td align="left">const</td>
</tr>
<tr class="even">
<td align="left">binomial</td>
<td align="left">logit</td>
<td align="left">np(1-p)</td>
</tr>
<tr class="odd">
<td align="left">poisson</td>
<td align="left">log</td>
<td align="left">mu</td>
</tr>
<tr class="even">
<td align="left">gamma</td>
<td align="left">reciprocal</td>
<td align="left">mu^2</td>
</tr>
<tr class="odd">
<td align="left">negbin</td>
<td align="left">log</td>
<td align="left">mu+mu^2/theta</td>
</tr>
</tbody>
</table>
<p>DESeq2 and edgeR use the negative binomial GLM for inference on differential
expression in RNA-seq. You will hear much more about this. The <code>glm.nb</code> function
in the MASS package allows you to fit negative binomial regressions with
data in data.frames.</p>
</div>
</div>
<div class="section level2">
<h2 id="hierarchical-models-alternatives-to-independence">Hierarchical models: alternatives to independence<a class="anchor" aria-label="anchor" href="#hierarchical-models-alternatives-to-independence"></a>
</h2>
<p>A very strong assumption in the use of linear and generalized linear
models is the mutual independence of reponses. When clustering
of responses is present (for example, when analyzing
data collected on human families, or pairs of eyes, or mice in litters), additional model structure
should be introduced to accommodate lack of independence.</p>
<div class="section level3">
<h3 id="random-effects-models-for-clustered-observations">Random effects models for clustered observations<a class="anchor" aria-label="anchor" href="#random-effects-models-for-clustered-observations"></a>
</h3>
<p>A framework approach developed in <a href="https://pubmed.ncbi.nlm.nih.gov/7168798/" class="external-link">1982 (Laird and Ware)</a>
expands the linear model to a model known as “mixed effects”:</p>
<p><span class="math display">\[
y_{ij} = \alpha + \beta x_{ij} + a_i + e_{ij}
\]</span></p>
<p>with <span class="math inline">\(i\)</span> indexing clusters and <span class="math inline">\(j\)</span> responses within clusters. The “random effect”
<span class="math inline">\(a_i\)</span> is assumed to follow <span class="math inline">\(N(0, \sigma^2_b)\)</span> (between-cluster variance), and
is independent of <span class="math inline">\(e_{ij} \sim N(0, \sigma^2_w)\)</span>.</p>
<p>The intraclass correlation coefficient measures the departure from independence, and
has form <span class="math inline">\(\rho = \sigma^2_b/(\sigma^2_b + \sigma^2_w)\)</span>.</p>
<p>Approximate mixed effects models for the GLM families can be fit using glmmPQL in MASS.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vince Carey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
